File: Algorithms/Dynamic_Programming/max_subarray.py
#!/usr/local/bin/python3

numCase = int(input())
for _ in range(numCase):
    arr_length = int(input())
    arr = list(map(int, input().strip().split(" ")))
    #print(arr)
    if max(arr) < 0:
        max_contig = max(arr)
        max_non_contig = max(arr)
        print(str(max_contig) + " " + str(max_non_contig))
    else:
        max_contig = 0
        max_contig_here = 0
        max_non_contig = 0
        for number in arr:
            #print(number)
            if number > 0:
                max_non_contig += number
            if max_contig_here + number > 0:
                max_contig_here = max_contig_here + number
                max_contig = max(max_contig_here, max_contig)
                #print(max_contig)
            else:
                max_contig_here = 0
        print(str(max_contig) + " " + str(max_non_contig))

File: Algorithms/Dynamic_Programming/MaxSubarray.java
import java.io.*;
import java.util.*;

public class MaxSubarray {

    public static void main(String[] args) {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution. */
        Scanner in = new Scanner(System.in);
        int numCase = in.nextInt();;
        for(int i = 0; i<numCase; i++){
            int arrLength = in.nextInt();
            int[] arr = new int[arrLength];
            int maxNonContig = 0;
            int maxContig = 0;
            int maxContigHere = 0;
            int arrMax = Integer.MIN_VALUE;
            for(int j = 0; j<arrLength; j++){
                arr[j] = in.nextInt();
                if (arr[j] > arrMax){
                    arrMax = arr[j];
                }
                if (arr[j] > 0) {
                    maxNonContig += arr[j];
                }
                if (maxContigHere + arr[j] > 0){
                    maxContigHere = maxContigHere + arr[j];
                    maxContig = Math.max(maxContigHere, maxContig);
                }
                else{
                    maxContigHere = 0;
                }
            }
            if(arrMax < 0){
                maxContig = arrMax;
                maxNonContig = arrMax;
            }
            System.out.println(maxContig + " " + maxNonContig);
        }
    }
}
